Add support for backing up MariaDB compressed and encrypted

janlindstrom opened the pull request:
tablespaces and tables. Furhermore, add support for backing
up MariaDB encrypted redo-logs.

janlindstrom committed:
Add support for backing up MariaDB compressed and encrypted
tablespaces and tables. Furhermore, add support for backing
up MariaDB encrypted redo-logs.

gl-sergei commented:
Hi,

Here is what I found:
- there is a memory leak in mariadb_check_tablespace_encryption
- patch won't compile on 32-bit platforms with following error: `mariadb.cc:103: error: format ‘%lu’ expects type ‘long unsigned int’, but argument 2 has type ‘ib_uint64_t’`

Backing up of encrypted and compressed tables seems working. Prepare doesn't work. Will it be separate patch? I didn't check backing up with redo log encryption enabled yet.

janlindstrom committed:
Fixed memory leak on mariadb_check_tablespace_encryption function and
fixed 32-bit platform compiler error mariadb.cc:103: error: format ‘%lu’
expects type ‘long unsigned int’, but argument 2 has type ‘ib_uint64_t’.

janlindstrom commented:
Hi,

Thank you for you review. Fixed issues on

https://github.com/janlindstrom/percona-xtrabackup/commit/0ad6809fc51f56b1fb4d188916b1eb60b172d2f7

Prepare will be fixed on MariaDB server with separate patch.

gl-sergei commented:
Just an update. I merged this patch to 2.4 (which is based on 5.7) and it seems to work OK, except one debug assertion when page_size is 8k:

```
2016-07-05 08:02:36 0x7f3b5a28e740  InnoDB: Assertion failure in thread 139892892428096 in file fsp0space.h line 141
InnoDB: Failing assertion: fsp_flags_is_valid(fsp_flags)
```

I had no time to look at it yet.

gl-sergei commented on storage/innobase/include/fsp0fsp.ic:
```diff
@@ -63,9 +63,20 @@ fsp_flags_is_valid(
 
+	/* If there is flags that are MariaDB extended flags,
+	read unused and page size from MariaDB positions. */
+	if ((unused && mariadb_unused == 0) ||
```
I think `unused != 0 && mariadb_unused == 0` would be more readable

janlindstrom replied:
Fixed.

gl-sergei commented on storage/innobase/include/fsp0fsp.ic:
```diff
@@ -63,9 +63,20 @@ fsp_flags_is_valid(
+	/* If there is flags that are MariaDB extended flags,
+	read unused and page size from MariaDB positions. */
+	if ((unused && mariadb_unused == 0) ||
+	    (page_ssize && (page_ssize < UNIV_PAGE_SIZE_MIN ||
```
same for `page_ssize != 0`

janlindstrom replied:
Fixed.

gl-sergei commented on storage/innobase/include/fsp0fsp.ic:
```diff
@@ -63,9 +63,20 @@ fsp_flags_is_valid(
+	read unused and page size from MariaDB positions. */
+	if ((unused && mariadb_unused == 0) ||
+	    (page_ssize && (page_ssize < UNIV_PAGE_SIZE_MIN ||
+	     page_ssize > UNIV_PAGE_SIZE_MAX))) {
```
Shouldn't it be `(page_ssize < UNIV_PAGE_SSIZE_MIN || page_ssize > UNIV_PAGE_SSIZE_MAX)` instead of `(page_ssize < UNIV_PAGE_SIZE_MIN || page_ssize > UNIV_PAGE_SIZE_MAX)`?

janlindstrom replied:
There is no UNIV_PAGE_SSIZE_MIN, but making this more clear.

gl-sergei commented:
Should

``` c++
/********************************************************************//**
Extract the page size from tablespace flags.
@return page size of the tablespace in bytes */
UNIV_INLINE
ulint
fsp_flags_get_page_size(
/*====================*/
    ulint   flags); /*!< in: tablespace flags */
```

also be updated?

However, I am not sure if xtrabackup using it during the copy.

Xtrabackup using `FSP_FLAGS_GET_ZIP_SSIZE` to determine page size. Does this field have the same position in MariaDB?

janlindstrom commented:
FSP_FLAGS_ZIP_SSIZE position is not changed.

janlindstrom commented:
I used latest Percona server 5.5 and configured it to use 8K page size, then tried to use 2.3 xtrabackup and it also fails on fsp_flags_is_valid() because:

```
  } else if (UNIV_PAGE_SIZE != UNIV_PAGE_SIZE_ORIG && !page_ssize) {
            fprintf(stderr, "InnoDB: Error: Tablespace flags %lu corrupted page_ssize %lu max %lu:%d\n",
                    flags, page_ssize, UNIV_PAGE_SIZE, UNIV_PAGE_SIZE_ORIG);
            return(false);
```

Looking into code, I do not see and ha_innodb::create ever storing page_ssize even if it is != UNIV_PAGE_SIZE_ORIG.
        }

janlindstrom committed:
After review fixes.

janlindstrom commented:
Percona Server 5.6 has following in fil0fil.cc:

```
/* Add the UNIV_PAGE_SIZE to the table flags and write them to the
tablespace header. */
flags = fsp_flags_set_page_size(flags, UNIV_PAGE_SIZE);
```

janlindstrom committed:
Fix error on determining page ssize when MariaDB extended tablespace
flags is used.

janlindstrom commented:
Found and fixed the error on page ssize determination when MariaDB flags were used.

io41 commented:
Is there any plans to update this PR @janlindstrom ?
@gl-sergei if this gets updated against master, are you willing to merge it?

Thanks!

janlindstrom commented:
Removing this PR due to (1) it is based on old MariaDB version and (2) no actual progress.

io41 commented:
In case anyone else finds this PR for the same reason I did, here's where the train is going: https://mariadb.com/resources/blog/mariadb-backup-released-mariadb-server-10123

hydrapolic commented:
Thanks @io41 for the update, haven't noticed in the changelog.

This will be tricky (https://jira.mariadb.org/browse/MDEV-9566):
Missing:
    Update xtrabackup sources (not sure how to do that easily and repeatable)
